vscode/DscExt/src/extension.ts

```diff
@@ -0,0 +1,67 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import * as fs from 'fs';
+
+export function activate(context: vscode.ExtensionContext) {
+
+	const dsc_type_provider = vscode.languages.registerCompletionItemProvider({ language: 'yaml', pattern: '**/*.dsc.yaml' },
+		{
+			provideCompletionItems(document: vscode.TextDocument, position: vscode.Position) {
+
+				const linePrefix = document.lineAt(position).text.slice(0, position.character);
+				if (!linePrefix.endsWith('type:')) {
+					return undefined;
+				}
+
+				let ps_cache_file_path: string = '';
+				if (process.platform==='win32')
+				{
+					ps_cache_file_path = process.env.LocalAppData + '\\dsc\\PSAdapterCache.json';
+				}
+				else
+				{
+					ps_cache_file_path = process.env.HOME + '/.dsc/PSAdapterCache.json';
+				}
+				console.log("Using cache path " + ps_cache_file_path);
+
+				if (!fs.existsSync(ps_cache_file_path)) {
+					console.log("Cache file does not exist");
+					return [];
+				}
+
+				const dataArray = JSON.parse(fs.readFileSync(ps_cache_file_path, 'utf-8'));
+				var completions:vscode.CompletionItem[] = [];
+				dataArray.ResourceCache.forEach(function (resouce: any) {
```

Comment:
Just noting `resouce` dropped an `r` 😅

Reply:
It's just an optimization but since (to my eye) everything in the lambda is...I don't know, is the word stateless? Not quite idempotent? But like the closure doesn't mutate anything, so we could fire all these completion builders off at once with `await Promise.all(dataArray.ResourceCache.map(async (resource: any)...` right?

Reply:
Ok so technically you're pushing onto `completions` from outside the closure but that was just an implementation choice, since this could also just return the completions from each async lambda and join them together as the final set of results. Is there any particular sorting we're wanting to have happen?